### PR TITLE
[TOOLS-2927] Scenarios Dashboard - Logbook - Scenario

### PR DESCRIPTION
**Pull request by jytakano**

### CODE REVIEW

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioSectionGenerator.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                        , "100%"),
+                asList("Weekly Throughput"
+                        , formatHourValue(plannedWeeklyThroughput)
+                        , formatHourValue(actualWeeklyThroughput)
+                        , formatPercentage(accuracyWeeklyThroughput)));
+
+        return new LogbookSectionTableDto(title, headers, rows);
+    }
+
+    private BigDecimal calculateWeekWorklog(LocalDate weekDate) {
```
Comment (taksan) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioSectionGenerator.java]: I would create a method in FollowupSnapshot that's like this:
```
public double cumulativeWorklog(LocalDate from, LocalDate to) {
        return followupData.fromJiraDs.rows.stream()
            .flatMap(row -> row.worklogs.stream())
            .filter(w -> Range.between(from, to).contains(toLocalDate(w.started, timezone)))
            .mapToDouble(w->w.timeSpentSeconds)
            .sum() / 3600;
```

File: strategical-dashboard/src/test/java/objective/taskboard/monitor/ScenarioSectionGeneratorTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                logbook()
+                    .withWeekDate("2019-03-18")
+                    .withWeekParameters(
+                        weekParameters()
+                            .withWorkload(3.0)
+                            .withEfficiency(0.67)
+                            .withWeeklyCapacity(35)
+                            .withWeeklyThroughput(70.35)
+                            .build()))
+            .cumulativeWorklogForDate("2019-03-03", 0)
```
Comment (taksan) [inline strategical-dashboard/src/test/java/objective/taskboard/monitor/ScenarioSectionGeneratorTest.java]:
```
.cumulativeWorklogs()
  at("2019-03-03").value(0)
  at("2019-03-17").value(200)
.scopeDone()
  .at("2019-03-03").value(0)
  .at("2019-03-10").value(110)
  .at("2019-03-17").value(200)
```

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScopeSectionGenerator.java [hunk 3]
```diff
@@ -77,15 +75,23 @@ public RowBuilder(LogbookModel logbook, ScopeChartData scopeData) {
             BigDecimal cumulativeAccuracy = calculateAccuracy(cumulativeEstimated, cumulativeActual);
 
             String cumulativeRange = getOrClosest(topOnTrackMap, logbookWeekDate)
-                    .map(upper -> removeDecimal(upper, "h"))
+                    .map(upper -> formatHour(upper))
                     .flatMap(upper -> getOrClosest(bottomOnTrackMap, logbookWeekDate)
-                            .map(lower -> removeDecimal(lower, "h"))
-                            .map(lower -> lower + "-" + upper))
-                    .orElse("");
+                            .map(lower -> formatHour(lower))
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScopeSectionGenerator.java]: MINOR Replace this lambda with a method reference. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1612)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScopeSectionGenerator.java [hunk 4]
```diff
@@ -77,15 +75,23 @@ public RowBuilder(LogbookModel logbook, ScopeChartData scopeData) {
             BigDecimal cumulativeAccuracy = calculateAccuracy(cumulativeEstimated, cumulativeActual);
 
             String cumulativeRange = getOrClosest(topOnTrackMap, logbookWeekDate)
-                    .map(upper -> removeDecimal(upper, "h"))
+                    .map(upper -> formatHour(upper))
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScopeSectionGenerator.java]: MINOR Replace this lambda with a method reference. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1612)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/LogbookSectionUtils.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package objective.taskboard.monitor;
+
+import static objective.taskboard.utils.NumberUtils.numberEquals;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.text.NumberFormat;
+import java.util.Locale;
+
+public class LogbookSectionUtils {
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/LogbookSectionUtils.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1118)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioSectionGenerator.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            weekWorklog.divide(weeklyCapacity, 1, RoundingMode.HALF_UP);
+    }
+
+    private BigDecimal calculateActualEfficiency(BigDecimal weekScopeDone, BigDecimal weekWorklog) {
+        return numberEquals(weekWorklog, BigDecimal.ZERO) ?
+            BigDecimal.ZERO :
+            weekScopeDone.divide(weekWorklog, 3, RoundingMode.HALF_UP);
+    }
+
+    private BigDecimal calculateWeekScopeDone(LocalDate weekDate) {
```
Comment (taksan) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioSectionGenerator.java]: I would have this logic inside FollowupSnapshotImpl, like
`scopeDoneBetween(from, to)`

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioSectionGenerator.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        return BigDecimal.valueOf(weekWorklog);
+    }
+
+    private BigDecimal calculateActualWorkload(BigDecimal weekWorklog, BigDecimal weeklyCapacity) {
+        return numberEquals(weeklyCapacity, BigDecimal.ZERO) ?
+            BigDecimal.ZERO :
+            weekWorklog.divide(weeklyCapacity, 1, RoundingMode.HALF_UP);
+    }
+
+    private BigDecimal calculateActualEfficiency(BigDecimal weekScopeDone, BigDecimal weekWorklog) {
```
Comment (taksan) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioSectionGenerator.java]: If `weekScopeDone == 0` and `weekWorklog == 0`, what is the efficiency?

**Pull request merged**